haskell-auto-import.el

```diff
@@ -0,0 +1,286 @@
+     (reverse (cdr parts)))))
+
+(defun haskell-auto-import-module-single-import-info (module-name &optional start)
+  "Returns information about the first import of the given module after `start'
+(which is set to 0 if nil). This information is nil if the module is not
+imported. Otherwise it is a list consisting of `(list start-point end-point
+is-qualified qualified-name imported-items)' where:
+- `start-point' is the point at the start of the import line
+- `end-point' is the point at the end of the import line
+- `is-qualified' is `t' if the import is qualified, nil otherwise
+- `qualified-name' is nil if no `as' clause was specified, or the qualified name
+  otherwise.
+- `imported-items' is a list of items imported from the module, in the format
+  provided by `haskell-auto-import-parse-import-parens' if it exists, nil otherwise."
+  (let ((contents (buffer-string))
+        (regexp
+         (concat
+          ;; Match the import keyword followed by whitespace
+          "^import\\s-+"
+          ;; Optionally match the qualified keyword as group 1
+          "\\(qualified\\s-+\\)?"
+          ;; Match exactly the module name
+          (regexp-quote module-name)
+          ;; Optionally match an `as' qualifier as group 2
+          "\\(?:\\s-+as\\s-+\\(\\sw+\\)\\)?"
+          ;; Whitespace may or may not precede parentheses at the end
+          "\\s-*"
+          ;; Optionally match the contents of some following parentheses as
+          ;; group 3
+          "\\(?:(\\(\\(?:[a-zA-Z0-9', \t\f\n\r]*\\|(.*?)\\)*\\))\\)?"
+          )))
+    (if (string-match regexp contents start)
+        (let ((is-qualified (match-string 1 contents))
+              (qualified-name (match-string 2 contents))
+              (imported-items-str (match-string 3 contents))
+              (start-point (match-beginning 0))
+              (end-point (match-end 0)))
+          (list start-point end-point
+                (if is-qualified t nil)
+                qualified-name
+                (if imported-items-str
+                    (haskell-auto-import-parse-import-parens imported-items-str)
+                  nil)
+                )))))
+
+(defun haskell-auto-import-module-import-info (module-name &optional start)
+  "Returns information about every import of the given module after `start'
+(which is set to 0 if nil). Each member of the returned list has the same format
+as that returned from `haskell-auto-import-module-single-import-info'."
+  (let ((next-info (haskell-auto-import-module-single-import-info module-name start)))
+    (if next-info
+        (let ((next-start (car (cdr next-info))))
+          (cons next-info (haskell-auto-import-module-import-info module-name next-start)))
+      ()
+      )))
+
+(defun haskell-auto-import-normalize-imported-items (imported-items)
+  "Normalize imported items, e.g. `one(a), two(a, b), two(a, c), one' --> `one,
+two(a, b, c)'"
+  (let ((grouped (-group-by 'car imported-items)))
```

Comment:
I notice that haskell-mode doesn't use dash, so you'll have to stick to the built-ins.